File: labsync-lms-backend/src/main/java/com/labsync/lms/controller/ScheduleController.java
package com.labsync.lms.controller;

import com.labsync.lms.dto.request.ScheduleRequest;
import com.labsync.lms.dto.response.ApiResponse;
import com.labsync.lms.dto.response.ClashReport;
import com.labsync.lms.dto.response.ScheduleResponse;
import com.labsync.lms.model.Schedule;
import com.labsync.lms.service.ScheduleService;
import jakarta.validation.Valid;
import org.springframework.http.*;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import java.util.List;

/**
 * ScheduleController — manual schedule CRUD + clash validation endpoints.
 *
 * Base URL: /schedules
 *
 * Key endpoints:
 *   POST /schedules/validate  — check for clashes without creating
 *   POST /schedules           — create a schedule (validates automatically)
 *   GET  /schedules/batch/{id} — get all sessions for a batch
 *   GET  /schedules/day/{id}   — get all sessions on a day
 *
 * Sample create request:
 * {
 *   "batchId": 1,
 *   "subjectId": 2,
 *   "staffId": 1,
 *   "labId": 1,
 *   "dayId": 1,
 *   "slotId": 2
 * }
 *
 * Sample clash response (409):
 * {
 *   "success": false,
 *   "message": "Schedule clash detected",
 *   "data": {
 *     "clashes": [
 *       "Lab 'Lab-101' is already booked on Monday at Slot-2",
 *       "Staff 'Dr. Patil' is already assigned on Monday at Slot-2"
 *     ]
 *   }
 * }
 */
@RestController
@RequestMapping("/schedules")
public class ScheduleController {
    private final ScheduleService scheduleService;

    // ── Clash Validation (dry-run) ─────────────────────────────────────────────
    /**
     * POST /schedules/validate
     * Validates a schedule request without persisting — returns all clashes.
     * Useful for frontend pre-validation before submission.
     */
    @PostMapping("/validate")
    public ResponseEntity<ApiResponse<ClashReport>> validateSchedule(@Valid @RequestBody ScheduleRequest request) {
        ClashReport report = scheduleService.validateSchedule(request);
        String message = report.isHasClash() ? "Clashes detected — schedule cannot be created" : "No clashes detected — schedule is valid";
        return ResponseEntity.ok(ApiResponse.success(message, report));
    }

    // ── CRUD ──────────────────────────────────────────────────────────────────
    @GetMapping
    public ResponseEntity<ApiResponse<List<ScheduleResponse>>> getAllSchedules() {
        return ResponseEntity.ok(ApiResponse.success("All schedules fetched", scheduleService.getAllSchedules()));
    }

    @GetMapping("/public")
    public ResponseEntity<ApiResponse<List<ScheduleResponse>>> getPublicSchedules() {
        return ResponseEntity.ok(ApiResponse.success("Public schedules fetched", scheduleService.getAllSchedules()));
    }

    @PostMapping
    @PreAuthorize("hasRole(\'ADMIN\')")
    public ResponseEntity<ApiResponse<ScheduleResponse>> createSchedule(@Valid @RequestBody ScheduleRequest request) {
        Schedule created = scheduleService.createSchedule(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success("Schedule created", scheduleService.toResponse(created)));
    }

    @PutMapping("/{id}")
    @PreAuthorize("hasRole(\'ADMIN\')")
    public ResponseEntity<ApiResponse<ScheduleResponse>> updateSchedule(@PathVariable Long id, @Valid @RequestBody ScheduleRequest request) {
        Schedule updated = scheduleService.updateSchedule(id, request);
        return ResponseEntity.ok(ApiResponse.success("Schedule updated", scheduleService.toResponse(updated)));
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole(\'ADMIN\')")
    public ResponseEntity<ApiResponse<Void>> deleteSchedule(@PathVariable Long id) {
        scheduleService.deleteSchedule(id);
        return ResponseEntity.ok(ApiResponse.success("Schedule deleted", null));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<ScheduleResponse>> getScheduleById(@PathVariable Long id) {
        Schedule s = scheduleService.getScheduleById(id);
        return ResponseEntity.ok(ApiResponse.success(scheduleService.toResponse(s)));
    }

    // ── Filtered views ─────────────────────────────────────────────────────────
    @GetMapping("/batch/{batchId}")
    public ResponseEntity<ApiResponse<List<ScheduleResponse>>> getByBatch(@PathVariable Long batchId) {
        return ResponseEntity.ok(ApiResponse.success(scheduleService.getSchedulesByBatch(batchId)));
    }

    @GetMapping("/day/{dayId}")
    public ResponseEntity<ApiResponse<List<ScheduleResponse>>> getByDay(@PathVariable Long dayId) {
        return ResponseEntity.ok(ApiResponse.success(scheduleService.getSchedulesByDay(dayId)));
    }

    @GetMapping("/lab/{labId}")
    public ResponseEntity<ApiResponse<List<ScheduleResponse>>> getByLab(@PathVariable Long labId) {
        return ResponseEntity.ok(ApiResponse.success(scheduleService.getSchedulesByLab(labId)));
    }

    public ScheduleController(final ScheduleService scheduleService) {
        this.scheduleService = scheduleService;
    }
}

File: labsync-node-backend/src/controllers/scheduleController.js
const Schedule = require('../models/Schedule');

const mapSchedule = (s) => ({
    scheduleId: s._id || s.id,
    batchId: s.batch ? (s.batch._id || s.batch.id) : null,
    batchName: s.batch ? `${s.batch.batchName} (${s.batch.division})` : 'Unknown',
    subjectId: s.subject ? (s.subject._id || s.subject.id) : null,
    subjectCode: s.subject ? s.subject.subjectCode : 'Unknown',
    staffId: s.staff ? (s.staff._id || s.staff.id) : null,
    staffName: s.staff ? s.staff.fullName : 'Not Assigned',
    labId: s.lab ? (s.lab._id || s.lab.id) : null,
    labName: s.lab ? s.lab.labName : 'Unknown',
    dayId: s.day ? (s.day._id || s.day.id) : null,
    dayName: s.day ? s.day.dayName : 'Unknown',
    startTime: s.startTime,
    endTime: s.endTime,
    generatedByGA: s.generatedByGA
});

exports.getPublicSchedules = async (req, res) => {
    try {
        const schedules = await Schedule.find({ active: true })
            .populate('batch subject staff lab day');
        
        res.status(200).json({
            success: true,
            message: "Public schedules fetched",
            data: schedules.map(mapSchedule)
        });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.getAllSchedules = async (req, res) => {
    try {
        const schedules = await Schedule.find()
            .populate('batch subject staff lab day');
        
        res.status(200).json({
            success: true,
            message: "All schedules fetched",
            data: schedules.map(mapSchedule)
        });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.getByBatch = async (req, res) => {
    try {
        const data = await Schedule.find({ batch: req.params.batchId }).populate('batch subject staff lab day');
        res.status(200).json({ success: true, message: 'Fetched successfully', data });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.getByDay = async (req, res) => {
    try {
        const data = await Schedule.find({ day: req.params.dayId }).populate('batch subject staff lab day');
        res.status(200).json({ success: true, message: 'Fetched successfully', data });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.getByLab = async (req, res) => {
    try {
        const data = await Schedule.find({ lab: req.params.labId }).populate('batch subject staff lab day');
        res.status(200).json({ success: true, message: 'Fetched successfully', data });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.validateSchedule = async (req, res) => {
    try {
        const report = { hasClash: false, clashes: [] };
        res.status(200).json({ success: true, message: 'No clashes detected — schedule is valid', data: report });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.create = async (req, res) => {
    try {
        const payload = {
            batch: req.body.batchId,
            subject: req.body.subjectId,
            staff: req.body.staffId || null,
            lab: req.body.labId,
            day: req.body.dayId,
            startTime: req.body.startTime,
            endTime: req.body.endTime,
        };
        const data = await Schedule.create(payload);
        res.status(201).json({ success: true, message: 'Created successfully', data });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};

exports.delete = async (req, res) => {
    try {
        const data = await Schedule.findByIdAndDelete(req.params.id);
        if (!data) return res.status(404).json({ success: false, message: 'Not found' });
        res.status(200).json({ success: true, message: 'Deleted successfully', data: null });
    } catch (error) {
        res.status(500).json({ success: false, message: error.message });
    }
};
